src/server/wesnothd/server.cpp

```diff
@@ -1538,12 +1644,105 @@ void server::handle_join_game(player_iterator player, simple_wml::node& join)
 	// send notification of changes to the game and user
 	simple_wml::document diff;
 	bool diff1 = make_change_diff(*games_and_users_list_.child("gamelist"), "gamelist", "game", g->changed_description(), diff);
-	bool diff2 = make_change_diff(games_and_users_list_.root(), nullptr, "user",
-		player->info().config_address(), diff);
+	bool diff2 = make_change_diff(games_and_users_list_.root(), nullptr, "user", player->info().config_address(), diff);
 
 	if(diff1 || diff2) {
 		send_to_lobby(diff);
 	}
+
+	// remove from any queues they may have joined
+	for(int q : player->info().get_queues()) {
+		queue_info& queue = queue_info_.at(q);
+		if(!queue.players_in_queue_.empty()) {
+			queue.players_in_queue_.erase(std::remove(queue.players_in_queue_.begin(), queue.players_in_queue_.end(), player->info().name()));
+			send_queue_update(queue);
+		}
+	}
+}
+
+void server::handle_join_server_queue(player_iterator p, simple_wml::node& data)
+{
+	int queue_id = data.attr("queue_id").to_int();
+
+	if(queue_info_.count(queue_id) == 0) {
+		ERR_SERVER << "player " << p->info().name() << " attempted to join non-existing server-side queue for scenario " << data.attr("queue_id");
+		return;
+	}
+
+	queue_info& queue = queue_info_.at(queue_id);
+
+	// if they're not already in the queue, add them
+	if(!utils::contains(queue.players_in_queue_, p->info().name())) {
```

Comment:
Same thing again here. Check for the error and return.

Reply:
Updated.